sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CoGroupByKeyLoadTest.java

```diff
@@ -92,22 +92,22 @@ void loadTest() throws IOException {
     Optional<SyntheticStep> syntheticStep = createStep(options.getStepOptions());
 
     PCollection<KV<byte[], byte[]>> input =
-        applyStepIfPresent(
-            pipeline.apply("Read input", SyntheticBoundedIO.readFrom(sourceOptions)),
-            "Synthetic step for input",
-            syntheticStep);
+        pipeline.apply("Read input", SyntheticBoundedIO.readFrom(sourceOptions));
+    input = input.apply("Collect start time metrics (input)", ParDo.of(runtimeMonitor));
```

Comment:
This unfortunately won't capture the time spent in the upstream source. This seems like a current limitation of the model since sources can only exist at the pipeline roots. With SplittableDoFn it will be easier to inject things at the beginning of the pipeline.

If its important to include all time including the source, another option would be to add hooks in `SyntheticBoundedIO` to execute counter collection code. i.e. something like:

```java
SyntheticBoundedIO
  .readFrom(sourceOptions)
  .withBeforeRead(() -> timeDistribution.update(System.currentTimeMillis()));
```

If it's not important to capture absolutely all timing, then it's fine to leave as-is; but perhaps leave a comment noting the limitation.

Reply:
I knew about the limitation but didn't think of such way of dealing with it. Great idea. Let me just leave a comment here and provide a PR that adds `withBeforeRead` and then extend the pr with the even better way.

Reply:
Sounds good to me, thanks.